scripts/prepare-test-ledger.sh

```diff
@@ -1,86 +1,158 @@
+REPLACE_TOP=${REPLACE_TOP:-}
+
+MAINNET_START='2021-03-17 00:00:00'
+now=$(date +%s)
+mainnet_start=$(date --date="$MAINNET_START" -u +%s)
+
+# Ledger prefix to use for structuring ledger
+LEDGER_PREFIX=${LEDGER_PREFIX:-staking-$(( (now-mainnet_start)/7140/180 ))}
 
 echo "Script assumes mainnet's start was at epoch 0 on 17th March 2021, if it's not the case, update the script please" >&2
+echo "Usage: $0 [-r|--replace-top] [-n|--norm] [-c|--delegation_cutoff $DELEGATEE_CUTOFF] [-b|--key-balance $KEY_BALANCE] [-p|--ledger-prefix $LEDGER_PREFIX] <BP key 1> <BP key 2> ... <BP key n>" >&2
+echo "Consider reading script's code for information on optional arguments" >&2
+
+##########################################################
+# Parse arguments
+##########################################################
+
+KEYS=()
+while [[ $# -gt 0 ]]; do
+  case $1 in
+    -r|--replace-top)
+      REPLACE_TOP=1; shift ;;
+    -n|--norm)
+      NORM=1; shift ;;
+    -c|--delegation-cutoff)
+      DELEGATEE_CUTOFF="$2"; shift; shift ;;
+    -b|--key-balance)
+      KEY_BALANCE="$2"; shift; shift ;;
+    -p|--ledger-prefix)
+      LEDGER_PREFIX="$2"; shift; shift ;;
+    -*|--*)
+      echo "Unknown option $1"; exit 1 ;;
+    *)
+      KEYS+=("$1") ; shift ;;
+  esac
+done
 
-if [[ $# -eq 0 ]]; then
-  echo "Usage: ./scripts/prepare-test-ledger.sh <BP key 1> <BP key 2> ... <BP key n>" >&2
+if [[ "$REPLACE_TOP" != "" ]] && [[ "$NORM" != "" ]]; then
+  echo "Can't use --norm and --replace-top at the same time" >&2
   exit 1
 fi
 
-num_keys=$#
-num_keys_and_ghost=$((num_keys + NUM_GHOST_KEYS))
+num_keys=${#KEYS[@]}
+if [[ $num_keys -eq 0 ]]; then
+  echo "No keys specified" >&2
+  exit 1
+fi
 
-now=$(date +%s)
-mainnet_start=$(date --date='2021-03-17 00:00:00' -u +%s)
-epoch_now=$(( (now-mainnet_start)/7140/180 ))
+##########################################################
+# Download ledger
+##########################################################
 
-ledger_file="$epoch_now.json"
+ledger_file="$LEDGER_PREFIX.json"
 
-echo "Current epoch: $epoch_now" >&2
+echo "Using ledger with prefix: $LEDGER_PREFIX" >&2
 
 if [[ ! -f "$ledger_file" ]]; then
-  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1&prefix=staking-$epoch_now"
-  ledger_url=$(curl "$ledgers_url" | jq -r .items[0].mediaLink)
-
-  wget -O $ledger_file "$ledger_url"
+  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1000&prefix=$LEDGER_PREFIX"
+  echo "$ledgers_url" >&2
+  ledger_url=$(curl "$ledgers_url" | jq -r '.items | sort_by(.size|tonumber) | last.mediaLink')
+  if [[ "$ledger_url" == null ]]; then
+    echo "Couldn't fine ledger with prefix $LEDGER_PREFIX" >&2
+    exit 2
+  fi
+  curl "$ledger_url" >"$ledger_file"
+  cmp "$ledger_file" >/dev/null <<< "Ledger not found: next staking ledger is not finalized yet" && echo "Next ledger not finalized yet" >&2 && rm "$ledger_file" && exit 2
```

Comment:
beware: OpenSSL ships a program called `cmp` that is entirely unrelated. it is in my path before this handy one from diffutils 😱